### PR TITLE
Add target defines for wasm/emscripten

### PR DESCRIPTION
**Pull request by Keno**
This is perhaps a bit of an overeager commit, given that most of the
code to support wasm is not yet upstream, but I think it's important
to get our terminology clear early. This attempts to define platform
defines in both our C code and at the Julia level. For background,
wasm is a general standard for secure execution of untrusted code
arising out of the Javascript world (many web browsers these days
ship wasm support in their Javascript engine). However, wasm is not
tied to JS, so it would be good to avoid hardcoding any such
assumptions into our platform defines (e.g. people are considering
wasm for kernel space vms, etc.). Emscripten on the other hand is
the original C-to-JS compiler, as well as corresponding runtime
libraries, simulatenously fulfulling the role of C library and
operating system kernel (for some features at least - e.g.
providing file system emulation). As such, my suggestion is the
following:

- Sys.OS gets set to `Emscripten`, which is also what the emulated
  `uname` syscall returns when emulated by Emscripten, so is consistent
  with our documentation for what that global means.
- A new `isjsvm` function can be used to determine if we're running
  inside a JavaScript VM (I'd expect this check to be quite common,
  e.g. to call out to the native RegEx engine).
- `isunix` is true for Emscripten, but I don't envision it being true
  generally for either `isjvm` platforms or wasm platforms.
- I didn't currently add a specific check for wasm, since I'd suspect
  this to be rather rare (e.g. only if you wanted to do something like
  emit inline wasm, or have a specific implementation for wasm targets).
  However, you can of course always check Sys.MACHINE for
  wasm32-*-* or wasm64-*-*. If it turns out to be useful, we can add it.

The C code follows the same reasoning and defines _OS_EMSCRIPTEN_ and
_CPU_WASM_. I didn't bother defining something for jsvm, since the changes
to the C code will have to be quite closely tied to specifics of
Emscripten.

cc @SimonDanisch @tshort @staticfloat for thoughts.

### CODE REVIEW

**Review by staticfloat — APPROVED** [review]

**Comment by Keno** [comment]
That didn't turn out quite the way I expected. Will fix manually.

File: base/sysinfo.jl [hunk 1]
```diff
@@ -377,7 +384,19 @@ See documentation in [Handling Operating System Variation](@ref).
 """
 isapple(os::Symbol) = (os === :Apple || os === :Darwin)
 
-for f in (:isunix, :islinux, :isbsd, :isapple, :iswindows, :isfreebsd, :isopenbsd, :isnetbsd, :isdragonfly)
+"""
+    Sys.isjsvm([os])
+
+Predicate for testing if Julia is running in a JavaScript VM (JSVM), including e.g.
+WebAssembly in a browser or JavaScript embedding.
+WASM in a Browser/JS embedding)
```
Comment (ararslan) [inline base/sysinfo.jl]:
```suggestion
```

Comment (ararslan) [reply base/sysinfo.jl]: Sorry, I guess this didn't get picked up by the previous suggestion. I'm not sure how to make them multi-line...

Comment (Keno) [reply base/sysinfo.jl]: It's not a Browser or Javascript embedding though, it's a JS embedding in the browser.

Comment (ararslan) [reply base/sysinfo.jl]: Okay, how about delete this line as in the suggestion and amend the line above it to be
```
WebAssembly in a JavaScript embedding in the browser.
```
?

Comment (Keno) [reply base/sysinfo.jl]: I'd say `including e.g. a WebAssembly JavaScript embedding in the browser`, because it's a WebAssembly embedding in JavaScript.

Comment (ararslan) [reply base/sysinfo.jl]: Okay, sounds good to me.

**Review by ararslan — APPROVED** [review]

File: base/sysinfo.jl [hunk 2]
```diff
@@ -377,7 +384,18 @@ See documentation in [Handling Operating System Variation](@ref).
...
+Predicate for testing if Julia is running in a JavaScript VM (JSVM),
+including e.g. a WebAssembly JavaScript embedding in a web browser.
+
+!!! compat "Julia 1.2"
+    This function requires at least Julia 1.2.
+"""
+isjsvm(os::Symbol) = (os === :Emscripten)
+
+for f in (:isunix, :islinux, :isbsd, :isapple, :iswindows, :isfreebsd, :isopenbsd, :isnetbsd, :isdragonfly, :isjsvm)
     @eval $f() = $(getfield(@__MODULE__, f)(KERNEL))
```
Comment (ararslan) [inline base/sysinfo.jl]: So for JavaScript VM embeddings, we have `Sys.KERNEL === :Emscripten`?

Comment (Keno) [reply base/sysinfo.jl]: Well, only if you're building with Emscripten, but there isn't really any other way to build libjulia for wasm at the moment.

**Pull request merged**